proposals/resource_cache.md

```diff
@@ -0,0 +1,271 @@
+    context:
+    - name: deployment_count
+      resourceCache:
+        group: "apps"
+        version: "v1"
+        kind: "deployments"
+        namespace: "staging"
+        jmesPath: "items[?metadata.name=='{{ request.object.metadata.name }}'] || `[]` | length(@)"
+```
+
+https://kyverno.io/policies/linkerd/require-linkerd-server/require-linkerd-server/
+
+```yaml
+    context:
+    - name: server_count
+      apiCall:
+        urlPath: "/apis/policy.linkerd.io/v1beta1/namespaces/{{request.namespace}}/servers"
+```
+
+can be converted to:
+
+```yaml
+    context:
+    - name: server_count
+      resourceCache:
+        group: "policy.linkerd.io"
+        version: "v1beta1"
+        kind: "servers"
+        namespace: "{{request.namespace}}"
+```
+
+https://kyverno.io/policies/linkerd/check-linkerd-authorizationpolicy/check-linkerd-authorizationpolicy/
+
+```yaml
+    context:
+    - name: servers
+      apiCall:
+        urlPath: "/apis/policy.linkerd.io/v1beta1/namespaces/{{request.namespace}}/servers"
+        jmesPath: "items[].metadata.name || `[]`"
+    - name: httproutes
+      apiCall:
+        urlPath: "/apis/policy.linkerd.io/v1beta1/namespaces/{{request.namespace}}/httproutes"
+        jmesPath: "items[].metadata.name || `[]`"
+```
+
+can be converted to:
+
+```yaml
+    context:
+    - name: servers
+      resourceCache:
+        group: "policy.linkerd.io"
+        version: "v1beta1"
+        kind: "servers"
+        namespace: "{{request.namespace}}"
+        jmesPath: "items[].metadata.name || `[]"
+```
+
+
+https://kyverno.io/policies/istio/require-authorizationpolicy/require-authorizationpolicy/
+
+```yaml
+    - name: allauthorizationpolicies
+      apiCall:
+        urlPath: "/apis/security.istio.io/v1beta1/authorizationpolicies"
+        jmesPath: "items[].metadata.namespace"
+
+```
+
+can be converted to:
+
+```yaml
+    context:
+    - name: allauthorizationpolicies
+      resourceCache:
+        group: "security.istio.io"
+        version: "v1beta1"
+        kind: "authorizationpolicies"
+        jmesPath: "items[].metadata.namespace"
+```
+
+
+https://kyverno.io/policies/other/rec-req/require-netpol/require-netpol/
+
+
+```yaml
+    - name: policies_count
+      apiCall:
+        urlPath: "/apis/networking.k8s.io/v1/namespaces/{{request.namespace}}/networkpolicies"
+        jmesPath: "items[?label_match(spec.podSelector.matchLabels, `{{request.object.spec.template.metadata.labels}}`)] | length(@)"
+```
+
+can be converted to:
+
+```yaml
+    context:
+    - name: policies_count
+      resourceCache:
+        group: "networking.k8s.io"
+        version: "v1"
+        kind: "networkpolicies"
+        namespace: "{{request.namespace}}"
+        jmesPath: "items[?label_match(spec.podSelector.matchLabels, `{{request.object.spec.template.metadata.labels}}`)] | length(@)"
+```
+
+# Drawbacks
+
+API calls do not leverage caching by default.
+
+If needed, we can add a separate caching mechanism for API calls in the future.
```

Comment:
Updating drawbacks of this approach

```suggestion
1. API calls do not leverage caching by default. If needed, we can add a separate caching mechanism for API calls in the future.
2. Using `cache.kyverno.io/enabled: "true"` can cause issues when users forgets to add them to the resource. 
   1. It is easy to miss a resource when adding the label. `apiCall` will return all the resources of the given type while `resourceCache` will return only those resources that have the label. In case of 1-10k resources. 
   2. Users might not want to have a kyverno specific label in all their resources across all namespaces.
3. For the usecase mentioned in [motivation](#motivation), we need to add resource to the cache when the policies is applied, otherwise, when the resource is applied for the first time, it will fail because of the timeout like it currently does. This will take away the abilities to have substitutions in `resourceCache` (e.g. `namespace: "{{request.namespace}}"`) and the `resourceCache` field will have to be static.
```